src/core/server/saved_objects/export/apply_export_hooks.ts

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { SavedObject } from '../../../types';
+import { KibanaRequest } from '../../http';
+import { SavedObjectsTypeExportHook, SavedObjectsExportContext } from './types';
+
+interface ApplyExportHooksOptions {
+  objects: SavedObject[];
+  request: KibanaRequest;
+  exportHooks: Record<string, SavedObjectsTypeExportHook>;
+}
+
+// TODO: doc + add tests.
+export const applyExportHooks = async ({
+  objects,
+  request,
+  exportHooks,
+}: ApplyExportHooksOptions): Promise<SavedObject[]> => {
+  const context = createContext(request);
+  const byType = splitByType(objects);
+
+  let finalObjects: SavedObject[] = [];
+  for (const [type, typeObjs] of Object.entries(byType)) {
+    const typeHook = exportHooks[type];
+    if (typeHook) {
+      finalObjects = [...finalObjects, ...(await typeHook(typeObjs, context))];
+    } else {
+      finalObjects = [...finalObjects, ...typeObjs];
+    }
+  }
+
+  return finalObjects;
+};
```

Comment:
This is the core of the feature

The hook is defined by 

```ts
export type SavedObjectsTypeExportHook = <T = unknown>(
  objects: Array<SavedObject<T>>,
  context: SavedObjectsExportContext
) => SavedObject[] | Promise<SavedObject[]>;
```

Which allows to both:
- update exported SOs and return updated versions
- add additional objects to the export

Note that is also allows to filter / exclude objects from the export. This is probably something we do not want, but it seems still alright, and limiting that with a more structured API would need a better structure than just returning a list of SO from the hook.

Does that look alright?

Reply:
When I think of a "hook" I think of a callback that gets called for every model/object, like a "pre-save hook". What do you think about calling this `SavedObjectsExportTransform` and `registerExportTransform`?

Reply:
> Note that is also allows to filter / exclude objects from the export. This is probably something we do not want

I do like the simplicity of this API shape. Instead of solving the filtering caveat with the shape of the return type, any reason we shouldn't prevent accidental filtering at runtime by verifying that all object IDs that were passed in were also in the array that was returned?

> When I think of a "hook" I think of a callback that gets called for every model/object, like a "pre-save hook". What do you think about calling this `SavedObjectsExportTransform` and `registerExportTransform`?

+1 on not naming this concept "hook". We've had requests in the past for on-save and on-delete hooks and I worry this would confuse developers.

Reply:
I agree. Will rename to `transform` instead